viper/parser/parser.py

```diff
@@ -700,71 +701,86 @@ def pack_logging_topics(event_id, args, expected_topics, context):
     return topics
 
 
-def pack_args_by_32(holder, maxlen, arg, typ, context, placeholder):
+def pack_args_by_32(holder, maxlen, arg, typ, context, placeholder, dynamic_offset_counter=None, datamem_start=None):
+    """
+    Copy necessary variables to pre-allocated memory section.
+
+    :param holder: Complete holder for all args
+    :param maxlen: Total length in bytes of the full arg section (static + dynamic).
+    :param arg: Current arg to pack
+    :param context: Context of arg
+    :param placeholder: Static placeholder for static argument part.
+    :param dynamic_offset_counter: position counter stored in static args.
+    :param dynamic_placeholder: pointer to current position in memory to write dynamic values to.
+    :param datamem_start: position where the whole datemem section starts.
+    """
+
     if isinstance(typ, BaseType):
         value = parse_expr(arg, context)
         value = base_type_conversion(value, value.typ, typ)
         holder.append(LLLnode.from_list(['mstore', placeholder, value], typ=typ, location='memory'))
     elif isinstance(typ, ByteArrayType):
         bytez = b''
-        # Bytes from Storage
-        if isinstance(arg, ast.Attribute) and arg.value.id == 'self':
-            stor_bytes = context.globals[arg.attr]
-            if stor_bytes.typ.maxlen > 32:
-                    raise TypeMismatchException("Can only log a maximum of 32 bytes at a time.")
-            arg2 = LLLnode.from_list(['sload', ['add', ['sha3_32', Expr(arg, context).lll_node], 1]], typ=BaseType(32))
-            holder, maxlen = pack_args_by_32(holder, maxlen, arg2, BaseType(32), context, context.new_placeholder(BaseType(32)))
-        # String literals
-        elif isinstance(arg, ast.Str):
+
+        if isinstance(arg, ast.Str):
             if len(arg.s) > typ.maxlen:
                 raise TypeMismatchException("Data input bytes are to big: %r %r" % (len(arg.s), typ))
             for c in arg.s:
                 if ord(c) >= 256:
                     raise InvalidLiteralException("Cannot insert special character %r into byte array" % c)
                 bytez += bytes([ord(c)])
-            bytez_length = len(bytez)
-            if len(bytez) > 32:
-                raise InvalidLiteralException("Can only log a maximum of 32 bytes at a time.")
-            holder.append(LLLnode.from_list(['mstore', placeholder, bytes_to_int(bytez + b'\x00' * (32 - bytez_length))], typ=typ, location='memory'))
-        # Variables
-        else:
-            value = parse_expr(arg, context)
-            if value.typ.maxlen > typ.maxlen:
-                raise TypeMismatchException("Data input bytes are to big: %r %r" % (value.typ, typ))
-            holder.append(LLLnode.from_list(['mstore', placeholder, ['mload', ['add', value, 32]]], typ=typ, location='memory'))
+
+        source_expr = Expr(arg, context)
+        if isinstance(arg, ast.Str):
```

Comment:
Does `if isinstance(arg, ast.Str)` need to be in there twice? (see 725)

Reply:
Will have a lookg thanks